### ISSUE TITLE
Feat: command default value

### ISSUE THREAD

**Issue by tunnckoCore** [lukeed/sade#32]
Allow passing default value to a command.

Currently you can only pass `opts.default: true` to make the command default to the whole CLI.

Problem comes from the

```js
if (opts.default) this.default = cmd;
```

Use case:

```js
sade
  .command('run [file]', 'Run your file', {
    default: 'src/index.js',
    // default: true,
    alias: ['r'],
  })
  .example('run')
  .example('run file.js')
  .action((file, argv) => {
    console.log(file); // currently `undefined`
  });
```

**Comment by tunnckoCore** [comment]
What I've done is

```js
// last one (`opts.default`) should be deprecated?
const cmdDefaults = opts.defaultCommand || opts.defaults || opts.default;
const defaultValues =
  cmdDefaults && typeof cmdDefaults !== 'boolean'
    ? [].concat(cmdDefaults).filter(Boolean)
    : [];

this.curr = cmd;
if (cmdDefaults === true) {
  this.settings.defaultCommand = cmd;
}

const cmdData = {
  usage,
  options: [],
  alias: {},
  default: {},
  examples: [],
  defaultValues,
};

this.tree[cmd] = cmdData;
this.commandAliases[cmd] = [].concat(opts.alias).filter(Boolean);
```

In the `command()`, with of course other changes needed. And it's backward compatible.

The thing is the following.

1. if you provide `default: true` or `defaults` (just for the typos sake), the command is a default one e.g. `yarn` instead of `yarn isntal`.
2. if `default(s)` is not a boolean, it's a default value for the command eg. `command('build <input> [out]', { defaults: 'src' })`
  - it can also accept array, so if you have a command like `command('build <src> <dist>')` you can set default values for both; `command('build <input> [out]', { default: ['src', 'dist'] })`
3. new option `defaultCommand` which is also can be accepted through `sade(name, options)`, so you can define what's your default cli command, from the beginning, `sade('hela', { defaultCommand: 'help' })` and later still have `default(s)` on per command basis.

Works brilliantly.

**Comment by lukeed** [comment]
Sorry, forgot to reply to this one.

I don't understand the use or reasoning behind this? We already have `opts.default` for a command to designate which command should be run in the absence of a command's name (`yarn install` on `yarn`).

And if you **are going to** provide a default argument value, then that argument **is not** a required argument. In which case, you should not be using `<>` symbols. That's the semantic/defining characteristics of `<>` vs `[]`

**Comment by tunnckoCore** [comment]
True. And yet you may still want to define aruments default values. And it's kind of better to define it there, than in the action's code with `||` or such. Action is like the business logic, the `sade`'s methods are pure definitions and more visually distinguishable and intuitive.

**Comment by lukeed** [comment]
Arguably, a default value _is_ business logic.

All my actions that have any defaults are defined [like this](https://github.com/lukeed/pwa/blob/master/packages/cli/lib/init.js#L165). I purposefully _want_ any required arguments (`type` in this case) to fail early, because that's the point of having something be required. 

I'm pretty sure I can't be swayed on this one, sorry. It's a pretty clear violation of the vocabulary (which exists far beyond JS world). Passing default values for a _required_ argument feels all kinds of wrong. And the arguments' positions are respected for parameterization _so that_ you can easily define defaults and test business logic.

```js
module.exports = function (type, dir='foobar', opts={}) {
 //
}
```

**Comment by tunnckoCore** [comment]
> It's a pretty clear violation of the vocabulary (which exists far beyond JS world). Passing default values for a required argument feels all kinds of wrong

It was example. And yes I agree. I switched to talk about the optional ones.

> (type, dir='foobar', opts={})

Yes, but not always possible, or acceptable, or likable. I, personally, some times don't like to use default params for a reason and I'm defining them in the body. But it's even about me here, it's about unknown third parties.

It's a great thing to think a bit ahead and not walk against the consumer/implementer/end-user/whtever. Such little things (like that that I'm kind of forced to make it `foo || 'bar'` and so impacts test coverage and so the developer/I need to cover it) driving me crazy some times. And I can bet it's not only me. 

Now, yes of course, we are not forced. Of course no one will make you a problem for one test branch not covered or whatever. Never will be a problem to be 99.70% or whatever. But little things compound. It's the main reasons ecosystems like JavaScript can be so huge and in the same time so broken and useless in the most part. It's not exactly because most of packages are one-off shots by some one-off developer or newbie. It's because decisions made by the rest.

It's a mindset. All this can look like super weird and like little things, but can be noticed and make a difference.

Sorry for the lecture, don't get me wrong, just throughts : ))

**Comment by lukeed** [comment]
> Sorry for the lecture, don't get me wrong, just spontaneous throughts : ))

All good! And not worry

> Never will be a problem to be 99.70% or whatever.

Just test the default behavior, lol

> some times don't like to use default params

Then define the default in the function :)

---

Sorry, I'm thoroughly unconvinced here. It really just reads as noise when constructing your CLI. And it encroaches on the territory (again) of "Why can't I have a default value for required parameters?" 

I'd like to have (and sure, enforce) strict areas of concern where you do certain things. 

I'm of the opinion that JS ecosystem is _too loose_/_too flexible_ the majority of times, and that's why it's hard to piece things together or figure things out. When utilities are big, confusing, unclear, ...etc, then you end up with useless tools that need replacing.

The TLDR for my answer is that this violates separation of concerns. Your handler needs to handle/react to incoming values. Required's are required. Optionals are optional. It's up to your handler how (or when or if) to respond to undefined optional arguments.

Hope that makes sense

**Comment by tunnckoCore** [comment]
Sure ;)

> I'm of the opinion that JS ecosystem is too loose/too flexible the majority of times, and that's why it's hard to piece things together or figure things out. When utilities are big, confusing, unclear, ...etc, then you end up with useless tools that need replacing.

Valid point too.

Hm. Just remember in the beginning (around 2013) and a few following years I was exactly in that "strict" team. Seems like I loose (pun intended) that through the years and switched teams, haha.

**Comment by lukeed** [comment]
That, or this is just one of the few times you've found found yourself needing slightly more flexibility (even though there's no lack of functionality) 😉
